File: Estruturais/Facade.py
# Subsistemas complexos
class Amplifier:
    def on(self):
        print("Amplificador ligado.")
    
    def set_volume(self, volume):
        print(f"Volume ajustado para {volume}.")

class DvdPlayer:
    def on(self):
        print("DVD Player ligado.")
    
    def play_movie(self, movie):
        print(f"Reproduzindo o filme '{movie}'.")

class Projector:
    def on(self):
        print("Projetor ligado.")
    
    def set_input(self, source):
        print(f"Fonte de entrada do projetor definida como {source}.")

# Fachada
class HomeTheaterFacade:
    def __init__(self):
        self.amplifier = Amplifier()
        self.dvd_player = DvdPlayer()
        self.projector = Projector()

    def watch_movie(self, movie):
        print("Iniciando o sistema de home theater...")
        self.amplifier.on()
        self.amplifier.set_volume(10)
        self.dvd_player.on()
        self.dvd_player.play_movie(movie)
        self.projector.on()
        self.projector.set_input("DVD")
        print("Aproveite o filme!\n")

# Uso da Fachada
home_theater = HomeTheaterFacade()
home_theater.watch_movie("Inception")

File: Estruturais/Facade.java
package Estruturais;

// Subsistemas complexos
class Amplifier {
    public void on() {
        System.out.println("Amplificador ligado.");
    }
    
    public void setVolume(int volume) {
        System.out.println("Volume ajustado para " + volume + ".");
    }
}

class DvdPlayer {
    public void on() {
        System.out.println("DVD Player ligado.");
    }
    
    public void playMovie(String movie) {
        System.out.println("Reproduzindo o filme '" + movie + "'.");
    }
}

class Projector {
    public void on() {
        System.out.println("Projetor ligado.");
    }
    
    public void setInput(String source) {
        System.out.println("Fonte de entrada do projetor definida como " + source + ".");
    }
}

// Fachada
class HomeTheaterFacade {
    private Amplifier amplifier;
    private DvdPlayer dvdPlayer;
    private Projector projector;

    public HomeTheaterFacade() {
        this.amplifier = new Amplifier();
        this.dvdPlayer = new DvdPlayer();
        this.projector = new Projector();
    }

    public void watchMovie(String movie) {
        System.out.println("Iniciando o sistema de home theater...");
        amplifier.on();
        amplifier.setVolume(10);
        dvdPlayer.on();
        dvdPlayer.playMovie(movie);
        projector.on();
        projector.setInput("DVD");
        System.out.println("Aproveite o filme!\n");
    }
}

// Uso da Fachada
public class Facade {
    public static void main(String[] args) {
        HomeTheaterFacade homeTheater = new HomeTheaterFacade();
        homeTheater.watchMovie("Inception");
    }
}

